### PR TITLE
Create Configurable Metadata

### PR DESCRIPTION
**Pull request by lukemarkwordtlibertyits**
# Description

The Oauth-Proxy should have the ability to override an issuer's metadata if necessary.

```json
{
  "api_category": "/override_Endpoints",
  "upstream_issuer": "http://example.com/issuer",
  "manage_endpoint": "http://example.com/account",
  "custom_metadata": {
    "authorization_endpoint": "http://example.custom.com/authorization",
    "token_endpoint": "http://example.custom.com/token",
    "userinfo_endpoint": "http://example.custom.com/userinfo",
    "introspection_endpoint": "http://example.custom.com/introspect",
    "revocation_endpoint": "http://example.custom.com/revoke",
    "jwks_uri": "http://example.custom.com/keys",
    "issuer": "http://example.custom.com/keys",
  }
}
```

**GET Endpoints**

- User info | Used by Issuer Client
- Jwks | Used by Issuer Client

**POST Endpoints**

- authorization | Used by Issuer Client
- token | Used by Issuer Client
- introspect  | Used by Issuer Client
- revocation  | Used by Issuer Client
- issuer | Used by Issuer Client

~**Should the delete grants endpoint be included?**~

~The issuer with this endpoint, is that the oauth-proxy uses the okta-client to perform this logic. The delete grants endpoint for okta is specific to its own API. Other AuthZ servers may not contain the same delete grants endpoint.~

~If this functionality is required, then it may be worth creating a wrapper authz-client class.~

The Delete Grants endpoint is not specified on the OpenId spec, so it will not be taken into consideration.

# TODO

- [x] Create Scripted Tests for this functionality
- [x] Unit Tests
- [x] Implement

# Tests

Since the issuers that will need this functionality are not ready, testing this functionality will be a little tricky.

## JWKS endpoint

```sh
curl $HOST/oauth2/keys
```

**Expected**

- [x] Redirect to the proxied issuer's JWKS endpoint.

## User Info endpoint

```sh
curl $HOST/oauth2/userinfo
```

**Expected**

- [x] Redirect to the proxied issuer's User Info endpoint.

## Post Endpoint tests

- Go through auth follow
- Introspect against the default issuer
  - [x] should return false
- Introspect against the proxied issuer that is used in the introspect swap 
   - [x] should return true
- revoke
- Introspect against the proxied issuer that is used in the introspect swap
  - [x] should return false

## Regression Tests 
- [x] Pass with no changes to config
- [x] Pass when the custom_metadata values consistently point to a different issuer than the upstream_issuer

### CODE REVIEW

File: oauth-proxy/cli.js [hunk 1]
```diff
@@ -151,6 +151,53 @@ function processArgs() {
             required: false,
             default: null,
           },
+          custom_metadata: {
+            type: "array",
```
Comment (beaugrantham-va) [inline oauth-proxy/cli.js:155]: This isn't an array.

File: oauth-proxy/cli.js [hunk 2]
```diff
@@ -151,6 +151,53 @@ function processArgs() {
...
+            description:
+              "An array of endpoints that can be used to override the default metadata.",
+            required: true,
```
Comment (beaugrantham-va) [inline oauth-proxy/cli.js:158]: This should not be required.

**Comment by crolarlibertyva** [comment]
Works great so far! Was able to override all the endpoints to my local keycloak instance.

File: oauth-proxy/issuer_helper.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+const { Issuer } = require("openid-client");
+
+const buildIssuer = async (serviceConfig) => {
```
Comment (crolarlibertyva) [inline oauth-proxy/issuer_helper.js:3]: Should we add jsdocs here since it's a new module?

**Review by beaugrantham-va — APPROVED** [review]

**Review by crolarlibertyva — APPROVED** [review]

**Pull request merged**